### DISCUSSION TITLE
How and Where to Create Environment Variable?

### DISCUSSION THREAD

**Discussion by doncadavona** [coder/code-server#6321] [Q&A] [answered]
I usually run my code-server in the background via this command:

```
sudo systemctl enable --now code-server@$USER
```

Now, how and where do I add environment variables for it? I am also confused as to the configuration file and environment variable mentioned in the docs are the same or different.

The environment variable is mentioned multiple times in the docs:

```
https://coder.com/docs/code-server/latest
```

But there is no actual example of how and where to create it. Can we have an example `environment variable` file please?

**Comment by code-asher** [comment]
> Now, how and where do I add environment variables for it? 
 
You can add environment variables to a systemd service by editing 
the service file.  It supports both directly setting environment 
variables or setting a file that contains environment variables.

For example:

```
[Service]
Environment=Name=value
EnvironmentFile=/foo/env
```

> I am also confused as to the configuration file and environment 
> variable mentioned in the docs are the same or different.

Which environment variable are you referring to?  Some of them 
have equivalents in the config file (like `PASSWORD`) but most do 
not.  The environment variable takes precedence over the config 
file in cases where both are set.

**Comment by doncadavona** [comment]
I'm grasping some new knowledge on your comment @code-asher, I had no idea the docs were mentioning `systemd` configuration. It would be helpful if an example is made available. With some common configurations set:

I am trying to configure my code-server in many ways, such as changing the name to `something-else` from `code-server`. I find it hard to do that, I'm sure it was a new feature/option made: `--app-name`.

I am also trying to configure code-server to use my own proxy, to use something like `https://{{port}}.mydomain.com` instead of `https://mydomain.com/proxy/{{proxy}}`:

![image](https://github.com/coder/code-server/assets/8132015/10a778e8-647c-46a7-88be-5364ec3f1b77)

As of now, the configurations I mentioned are `VSCODE_PROXY_URI` and `APP_NAME`, if that exists.

**Reply by code-asher** [reply, marked as answer]
Ah yeah the way you set environment variables or add flags will depend on the way you run code-server, and it can also change depending on the operating system.  In your case that happens to be systemd and presumably GNU/Linux but there are other paths.

So with that in mind, maybe every time we talk about environment variables in the docs we should link to a section about environment variables and how to set them based on how you run code-server, although admittedly part of me thinks that how to set environment variables is a little too far out of scope for code-server's docs.

Alternatively, maybe instead of recommending environment variables we should recommend editing the config file and in cases where there is no equivalent config yet we add it (like the marketplace URL).  Then we will not have to worry about how code-server is being ran.

For `VSCODE_PROXY_URI`, using `--proxy-domain` should suffice so in your case I think I would edit the config file over editing the systemd file.  Here is an example config file:

```
bind-addr: 127.0.0.1:8080
auth: password
password: password
cert: false
proxy-domain: mydomain.com
app-name: something-else
```

If you are curious, the difference between using `--proxy-domain` and `VSCODE_PROXY_URI` is that `--proxy-domain`, in addition to setting the proxy URI, will also listen for requests to that domain and proxy them.  If you want to use code-server itself as a proxy then this is what you want anyway and if you are using an external proxy then it adds no measurable overhead so even though it is unused it should not be a problem.  But we should think about adding a new flag `--vscode-proxy-uri` or `--proxy-uri` to match the environment variable exactly.

If you need to configure something that cannot be set in the config file yet then here is an example systemd file:

```
[Unit]
Description=code-server
After=network.target

[Service]
Type=exec
Environment=VSCODE_PROXY_URI=https://{{port}}.mydomain.com
Environment=EXAMPLE=value
ExecStart=/usr/bin/code-server --app-name something-else
Restart=always

[Install]
WantedBy=multi-user.target 
```

**Reply by doncadavona** [reply]
That works! Not entirely as I was expecting, but it's good progress for me:

Adding some configurations to my `~/.config/code-server/config.yaml` and then restarting my `code-server` via `sudo systemctl restart code-server@$USER` works:

```
bind-addr: 127.0.0.1:8080
auth: password
password: SECRET
cert: false
proxy-domain: mydomain.com
app-name: something-else
```

Here are my findings so far:

- The `proxy-domain` configuration only prefixes the port to the value, for example, if the value is `mydomain.com`, then the resulting proxy will be `{{port}}.mydomain.com`. There is no way to place the port somewhere else. For example, `proxy-domain: subdomain-{{proxy}}.mydomain.com` does not work, because, it results to `{{port}}.subdomain-{{port}}.mydomain.com`.
- The `app-name: something-else` works, but it seems to show it only in the Welcome page, such as `code-server - something-else - login` and `Welcome to something-else` seen here:
![image](https://github.com/coder/code-server/assets/8132015/c817e09b-9495-46f9-8179-65f10ee772bf)
Upon login, there seems to be no app-name visible:
![Annotation 2023-07-19 085710](https://github.com/coder/code-server/assets/8132015/34bb7cd4-8e0a-424d-9f93-b0d56896375f)
- I tried setting the configuration to my `/lib/systemd/system/code-server@.service`, but right now they don't work.

By the way, I'm running `code-server v4.14.1` on a `Ubuntu 22.04` server.

**Reply by doncadavona** [reply]
I got it working now as I need:

I can add configurations to `~/.config/code-server/config.yaml` like so:
```
bind-addr: 127.0.0.1:8080
auth: password
password: password
cert: false
proxy-domain: mydomain.com
app-name: something-else
```

If I want to place the proxy somewhere else, like `subdomain-{{port}}.mydomain.com`, I have to add the configuration to `/lib/systemd/system/code-server@.service`:

```
[Unit]
Description=code-server
After=network.target

[Service]
Type=exec
Environment=VSCODE_PROXY_URI=https://subdomain-{{port}}.mydomain.com
Environment=EXAMPLE=value
ExecStart=/usr/bin/code-server --app-name something-else
Restart=always

[Install]
WantedBy=multi-user.target 
```

Then I need to run:
```
sudo systemctl daemon-reload
sudo systemctl restart code-server@$USER
```

Thank you @code-asher !

**Reply by code-asher** [reply]
Ah yeah the default behavior is to prefix but if `{{port}}` is included should let you do that.  `proxy-domain: subdomain-{{port}}.mydomain.com` should work.  For example (I removed the irrelevant parts):

```
code-server --proxy-domain subdomain-{{port}}.mydomain.com
[2023-07-19T17:33:55.618Z] info    - Proxying the following domain:
[2023-07-19T17:33:55.618Z] info      - subdomain-{{port}}.mydomain.com
[2023-07-19T17:33:55.619Z] info  Using proxy URI in PORTS tab: //subdomain-{{port}}.mydomain.com
```

I am not sure we actually have this documented yet though.

Good point on the app name, it seems to only affect the code-server wrapping parts and not VS Code.  I think one workaround is to edit the product name in `lib/vscode/product.json` although it is not a great workaround.
